Enhancement/issue 48

RHammond2 opened the pull request:
Closes #48 

This PR updates the QC methods to be bifurcated into a `QualityControlDiagnosticSuite` and `WindToolKitQualityControlDiagnosticSuite` to allow for a place to keep general QC methods and a way to extend the base method with case-specific QC methods. This PR also adds a more robust Daylight Savings Time handling through the use of UTC offsets instead of hard-coded date cutoffs.

@ejsimley  or @jordanperr the QC examples notebook seems to be breaking with the WTK data, do either of you have any insight on what's going on with this? I've only used WTK a couple of times and am unfamiliar with the specifics.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NREL/OpenOA?pullRequest=168) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/NREL/OpenOA?pullRequest=168) before we can accept your contribution.<br/><hr/>**Rob** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NREL/OpenOA?pullRequest=168) it.</sub>

ejsimley commented:
> the QC examples notebook seems to be breaking with the WTK data, do either of you have any insight on what's going on with this? I've only used WTK a couple of times and am unfamiliar with the specifics.

@RHammond2 I'm getting a WTK-related error running the example notebook too. I'm almost positive the error is caused by asking for WTK data for coordinates outside of the U.S. The WTK dataset we're using only covers the U.S. (or possibly North America?) but the example wind farm is in Europe. This is why `check_tz` was set to False in the example notebook previously.

RHammond2 commented:
@ejsimley I've updated the WIND Toolkit QC method to have both a `check_tz` parameter and to fail gracefully by passing a logging message that the coordinates are out of bounds. I've also added the pytz to the requirements per a discussion with Jordan, though pandas already has this as one of their own requirements.

I was unable to ammend the old commits in a way that would realign them with my GH account, so I think we can just ignore the CLA bit for that this time.

RHammond2 commented on operational_analysis/methods/quality_check_automation.py:
```diff
@@ -1,67 +1,181 @@
+    arr = np.array(
+        [
+            [True, pd.to_datetime(el).tz_localize(None).to_pydatetime()]
+            if not isinstance(el, float)
```
In theory yes, but in practice, this function is more geared toward internal data processing where at least in the issues i've faced having a float at this stage is where we run into trouble. I'll at least update the docstring so it can be more properly used in expansion efforts here.

RHammond2 commented on operational_analysis/methods/quality_check_automation.py:
```diff
@@ -1,67 +1,181 @@
+            self._df[self._t_local] = pd_dt_col
+        else:
+            pd_dt_col = pd.to_datetime(dt_col)
+            self._df[self._t_local] = pd_dt_col.tz_localize(self._local_tz, ambiguous=True)
```
I'll add an inline comment to the code, but this has to do with if the timestamps are timezone aware vs unaware and whether we should be converting the timestamp to local time, or if we should be localizing the timestamp. This flip in functionality from pandas is both a little confusing and mildly annoying, to be honest, so working around it is unfortunately similar, so I will clarify.

More directly though, the local time could entirely be UTC.

RHammond2 commented on operational_analysis/methods/quality_check_automation.py:
```diff
@@ -1,67 +1,181 @@
+            self._df = self._df.tz_convert("UTC")
+            self._df[self._t_utc] = self._df.index
+
+        self._df = self._determine_offset_dst(self._df)
```
This seems to be an error on my end where I need to set it back to local, or use a different method to update the field.

RHammond2 commented on operational_analysis/methods/quality_check_automation.py:
```diff
@@ -116,6 +215,328 @@ def gap_time_identification(self):
+        self._max_min["max"] = self._df.max()
+        self._max_min["min"] = self._df.min()
+
+    def _get_time_window(self, df, ix, hour_window):
```
I agree, I'll move that out of the class.

RHammond2 commented on operational_analysis/methods/quality_check_automation.py:
```diff
@@ -1,67 +1,181 @@
+
+    @abstractmethod
     @logged_method_call
     def run(self):
```
That makes sense, especially given recent v3 conversations on whether or not we should have users implement the functionality or us provide a way to use it more easily. I assume this would be all the same processes would run except the WTK features?

ejsimley replied:
yeah, exactly. Thanks for adding this.

RHammond2 commented on operational_analysis/methods/quality_check_automation.py:
```diff
@@ -116,6 +215,328 @@ def gap_time_identification(self):
+    """This class defines key analytical procedures in a quality check process for
+    turbine data. After analyzing the data for missing and duplicate timestamps,
+    timezones, Daylight Savings Time corrections, and extrema values, the user can make
+    informed decisions about how to handle the data.
```
I've added a general description, but let me know if you'd like to be more expansive or descriptive.

RHammond2 commented on operational_analysis/methods/quality_check_automation.py:
```diff
@@ -116,6 +215,328 @@ def gap_time_identification(self):
+            ix (:obj:`pandas._libs.tslibs.timestamps.Timestamp`]): The starting
+                Timestamp on which to base the time window.
+            hour_window (:obj:`pandas._libs.tslibs.timedeltas.Timedelta`): The number
+                length of the window, in hours.
```
I've updated the description to reflect this, so that the functionality is the same as the original, just more clear.

RHammond2 commented:
I've updated the following based on the notebook commentary:
 - The `_df` index is now called UTC as it should have been
 - The datetime column/index comment in 1.2 now makes sense and aligns with reality.
 - 1a.3.3 language has been updated for fall datetime duplications
 - Removed the confusing "Timezone-adjusted" language to simply be which ever timezone is being represented
 - Added commentary in 1a.3.4 regarding the duplicates that are still present in the spring data.
 - `dup_time_identification` and `gap_time_identification` now have a `which` keyword argument to allow customization for the timezone basis.
 - Updated to correct the wording that there is a time gap in the fall in 1b.3.3.
 - Found a bug in removing the timezones in the DST plots, so now the figures in notebooks 1A and 1B look highly similar, as expected.
 
 @ejsimley, let me know if there are still any lingering issues, or if any questions still remain in the above comments from your review.

ejsimley commented:
Rob, your latest changes definitely fix the confusion about the data frame index and the daylight savings time plots. I have a few last comments on the notebooks, which are listed here. 

- 1a.1.2: "The datetime column/index comment in 1.2 now makes sense and aligns with reality." Agreed, thanks for fixing that. But I think it's worth mentioning in the notebook that the datetime_utc column is calculated *assuming* that the original datetime is in local time, in case there are any questions about where the UTC value comes from when the datetime is tz-unaware.
- 1a.1.3.4: "Added commentary in 1a.3.4 regarding the duplicates that are still present in the spring data." I still think it's worth explaining why the spring plots don't show the duplicate data that was identified in 1.3.3. For example, duplicates were dropped before plotting the data.
- 1b.1.3.3: I'm not quite sure why example 1b (tz-aware example) identifies gaps in the fall, but the tz-unaware example 1a identifies gaps in the spring. Is this the intended behavior, or should this be fixed? If it is intended, then some discussion should be added explaining why the identified gaps are different in the two examples. 
- 1b.1.3.4: Minor point, but if it's not too much effort, it would be helpful to explicitly show the time gaps in the spring for the original timestamps (by plotting NaN values), instead of having the plot interpolate the gaps.

RHammond2 commented:
@ejsimley this last commit should take care of the remaining questions for both code changes and commentary. It took a bit longer than I thought it would, but I think it's a lot better for it once again, thanks!

jordanperr commented:
Triggering CI/CD pipeline

RHammond2 commented:
@ejsimley, the last commit should address the remaining discrepancies in the commentary. For 1b.1.3.3, there should not be a difference in the plots as of the last update other than a shift in the DST transition due to having the data accurately encoded. Let me know if this helps or not.

ejsimley commented on operational_analysis/methods/quality_check_automation.py:
```diff
@@ -1,120 +1,613 @@
-        ]
+        which = which.lower().strip()
+        col_map = dict(original=self._t, local=self._t_local, utc=self._t_utc)
+        which = "original" if which not in col_map else which
```
just noticed that "which" is no longer used in this function, right?

RHammond2 replied:
This was a leftover, I've just updated it to be like the duplication method, and added some commentary in the examples to highlight this functionality. Also, now that we're on the same page for the example commentary, this should hopefully be all set now!

ejsimley commented:
Hey Rob, the commentary on 1a.1.2 looks good. And sorry to keep bugging you about this, but for 1b.1.3.3, the plots look great. What I don't understand is why `qc._time_gaps` is showing time gaps in the Spring in example 1a but time gaps in the Fall for example 1b. I think this is somehow caused by how the data frame index is calculated for the tzaware vs. tzunaware cases but haven't fully figured it out. Is this the behavior you expect to see for the two cases? If so, I would suggest commenting on why different gaps are identified in the Fall for the two notebooks (yet the same duplicates are found). Or if it's not what you intended to happen, then maybe there is a bug somewhere?

ejsimley reviewed (approved):
I think this all looks good now. Thanks for your hard work on this! The additional commentary provides a nice guide through the notebooks. And I'm not too worried about the somewhat unexpected `qc._time_gaps` in example B since it's probably mostly tz-unaware data where someone would want to look at the time duplicates and gaps.
